feat(zeebe): add icon schema

pinussilvestrus opened the pull request:
Closes #50

Example 

```json
{
    "$schema": "https://unpkg.com/@camunda/zeebe-element-templates-json-schema/resources/schema.json",
    "name": "Rest Connector",
    "id": "icon.template.rest",
    "appliesTo": [
      "bpmn:ServiceTask"
    ],
    "properties": [],
    "icon": {

      // data URL
      "contents": "data:image/svg+xml,%3Csvg width='24' height='24' ... %3C/svg%3E"
      
      // simple URL
      "contents": "https://user-images.githubusercontent.com/58601/160609748-cc18d1cc-aa32-43d0-bac3-058a9979aa5b.svg"
    }
  }
```

nikku commented on packages/zeebe-element-templates-json-schema/test/spec/validationSpec.js:
```diff
@@ -189,6 +189,22 @@ describe('validation', function() {
+      testTemplate('icon');
+
+
+      testTemplate('icon-raw-svg');
```
Why do we also support this? Any particular reason?

nikku replied:
Raw SVG is a security concern; I#d like to not open that box.

nikku replied:
Images can just fine be embedded into SVG: https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Other_content_in_SVG#embedding_raster_images.

pinussilvestrus replied:
We do not support this, [the test case provides errors](https://github.com/camunda/element-templates-json-schema/blob/50-icon-schema/packages/zeebe-element-templates-json-schema/test/fixtures/icon-raw-svg.js#L13).

nikku replied:
Proof of works (IMG in SVG):

![foo](https://user-images.githubusercontent.com/58601/160609748-cc18d1cc-aa32-43d0-bac3-058a9979aa5b.svg)

nikku replied:
> We don't support this

Aaah, great. Maybe the test cases could be renamed for claritiy, i.e. "valid-xxx", "invalid-xxx".

pinussilvestrus replied:
Yes, I know. Still, raw SVG shouldn't be supported (what we are testing right now with this test), but only data URLs or `http(s):`

pinussilvestrus replied:
> Maybe the test cases could be renamed for claritiy, i.e. "valid-xxx", "invalid-xxx".
will do 👍

nikku commented on packages/zeebe-element-templates-json-schema/src/defs/icon.json:
```diff
@@ -0,0 +1,25 @@
+      "$id": "#/icon/contents",
+      "type": "string",
+      "description": "The svg contents of an icon.",
+      "pattern": "^(data:)(?<mime>[\\w/\\-\\.]+)",
```
I'm wondering why we don't settle down on:

* this can be a valid, safe `image#src` reference. This includes data-uri (including mime type, but also encoding) + relative or absolute http / https URIs.).

nikku replied:
We should account for encoding to be present, i.e.:

```
data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAIBAAA=
```

I'd love to see a relevant test.

But generally speaking: I'd be happy to relax the rules for the moment to just verify this is a `data:` or `http(s):` protocol URI.

pinussilvestrus replied:
Let's relax the pattern for the moment and just make sure that

> verify this is a data: or http(s): protocol URI.

pinussilvestrus commented:
Thanks for your feedback, I updated the test cases, added several more and adjusted the pattern: https://github.com/camunda/element-templates-json-schema/pull/51/commits/77c975e36f4679d8ec0a75a0a5413520e5f63bdf.

nikku commented on packages/zeebe-element-templates-json-schema/resources/schema.json:
```diff
@@ -649,6 +649,31 @@
+              "$id": "#/icon/contents",
+              "type": "string",
+              "description": "The URL of an icon.",
+              "pattern": "^((https|http)://|data:).*",
```
```suggestion
              "pattern": "^(https?|data):.*",
```

pinussilvestrus replied:
Will never get used to regex... It can be so easy 😄

nikku commented on packages/zeebe-element-templates-json-schema/src/defs/icon.json:
```diff
@@ -0,0 +1,25 @@
+  "description": "Custom icon to be shown on the element",
+  "default": {},
+  "properties": {
+    "contents": {
```
Given our direction we could rename this as `source`. But of course `contents` would do, too.

nikku reviewed (commented):
Very nice. Just two minor suggestions on error message + validation pattern.

pinussilvestrus commented on packages/zeebe-element-templates-json-schema/resources/schema.json:
```diff
@@ -649,6 +649,31 @@
+              "description": "The URL of an icon.",
+              "pattern": "^((https|http)://|data:).*",
+              "errorMessage": {
+                "pattern": "Malformed icon data URL, must match \"^((https|http)://|data:).*\""
```
Thanks! I updated it
